src/inputs/pcap/PcapInputStream.h

```diff
@@ -123,9 +146,45 @@ class PcapInputEventProxy: public visor::InputEventProxy
         packet_signal(payload, dir, l3, l4, stamp);
     }
 
+    // dns, <handlerid> by_qname2, google.com, <func that returns current qname>, <func that processes udp>
+    void register_udp_predicate_signal(const std::string &schema_key, const std::string &handler_id, const std::string &predicate_key, const std::string &conditional_value, UdpPredicate predicate, UdpSignalCB callback)
+    {
+        std::unique_lock lock(_pcap_proxy_mutex);
+        // if predicate has not been installed yet, install it now
+        // note that it is namespaced to each handler so that different handlers can have the same predicate_key
+        auto full_predicate_key = schema_key + predicate_key;
+        auto predicate_jump_key = full_predicate_key + conditional_value;
+        if (_udp_predicates.find(full_predicate_key) == _udp_predicates.end()) {
+            _udp_predicates[full_predicate_key] = predicate;
+        }
+        // now install the given conditional signal based on the jump key
+        // record the connection so we can remove it later when the handler disconnects
+        _udp_predicate_connections[handler_id] = _udp_predicate_signals[predicate_jump_key].connect(callback);
+    }
+
+    void unregister_udp_predicate_signal(const std::string &handler_id)
```

Comment:
My PR problem remains here:
If a policy that contains the one that caller `_udp_predicates` is removed, the others predicates will stop being called

Other point: we are not cleaning the `_udp_predicate_signals` map.

Reply:
hm i think you're right but only because i kept your code of the predicate being a member variable. i think in your code it had to be (?), but in this strategy it does not - it can be static because it's stateless (with the exception of the dns layer cache code)

we clean it by removing the connection, only other thing to do is remove the entry if there are 0 connections?

Reply:
hmm I think we could change it to static, we just need to ensure that it exists for the filter key right? like "by_qname2"

Reply:
actually, on second thought - `_udp_predicates` is never removed from, even if policies are removed so i think this is ok. it's just the signal connections which are cleaned up, not the predicate.